File: DesignPatternsExercises/java/InterpreterDesignPattern/Solution/ExpressionReader.java
public class ExpressionReader {
	// public __________(7)__________ Expression read(String expression) {
	public static Expression read(String expression) {
		String[] expressionToken = expression.split("\\s+");
		Constant firstConstant = new Constant(expressionToken[Statement.FIRST_CONSTANT_INDEX]);
		Constant secondConstant = new Constant(expressionToken[Statement.SECOND_CONSTANT_INDEX]);

		if (expression.contains(Statement.SUMMATION_OPERATOR))
			// return new __________(8)__________(firstConstant, secondConstant);
			return new SummationExpression(firstConstant, secondConstant);
		else if (expression.contains(Statement.SUBTRACTION_OPERATOR))
			// return new __________(9)__________(firstConstant, secondConstant);
			return new SubtractionExpression(firstConstant, secondConstant);
		else if (expression.contains(Statement.MULTIPLICATION_OPERATOR))
			// return new __________(10)__________(firstConstant, secondConstant);
			return new MultiplicationExpression(firstConstant, secondConstant);
		else if (expression.contains(Statement.DIVISION_OPERATOR))
			// return new __________(11)__________(firstConstant, secondConstant);
			return new DivisionExpression(firstConstant, secondConstant);
		return null;
	}
}

File: DesignPatternsExercises/ruby/interpreter_design_pattern/solution/expression_reader.rb
require_relative 'constant'
require_relative 'summation_expression'
require_relative 'subtraction_expression'
require_relative 'multiplication_expression'
require_relative 'division_expression'

class ExpressionReader
  # def __________(7)__________(expression)
  def self.read(expression)
    expression_token = expression.split(' ')
    first_constant = Constant.new(expression_token[Operators::FIRST_CONSTANT_INDEX])
    second_constant = Constant.new(expression_token[Operators::SECOND_CONSTANT_INDEX])

    # return __________(8)__________(first_constant, second_constant) if expression.include?(Operators::SUMMATION_OPERATOR.to_s)
    return SummationExpression.new(first_constant, second_constant) if expression.include?(Operators::SUMMATION_OPERATOR.to_s)
    # return __________(9)__________(first_constant, second_constant) if expression.include?(Operators::SUBTRACTION_OPERATOR.to_s)
    return SubtractionExpression.new(first_constant, second_constant) if expression.include?(Operators::SUBTRACTION_OPERATOR.to_s)
    # return __________(10)__________(first_constant, second_constant) if expression.include?(Operators::MULTIPLICATION_OPERATOR.to_s)
    return MultiplicationExpression.new(first_constant, second_constant) if expression.include?(Operators::MULTIPLICATION_OPERATOR.to_s)
    # return __________(11)__________(first_constant, second_constant) if expression.include?(Operators::DIVISION_OPERATOR.to_s)
    return DivisionExpression.new(first_constant, second_constant) if expression.include?(Operators::DIVISION_OPERATOR.to_s)

    nil
  end
end
